tests/unit/test_derivative_utils.py

```diff
@@ -1,27 +1,110 @@
+import httpx
 import pandas as pd
 import pytest
 
 from bagel import mappings
 from bagel.utilities import derivative_utils
 
 
+def test_get_pipeline_from_backup_if_remote_fails(monkeypatch):
+    """
+    Test that the pipeline catalog is loaded from the local backup if the remote location is unreachable.
+    """
+    # TODO: Make a proper mock for the requests.get function
+    # or switch to httpx for better testing capabilities
+    nonsense_url = "https://does.not.exist.url"
+
+    def mock_httpx_get(*args, **kwargs):
+        response = httpx.Response(status_code=400, json={"key": "value"})
+        # This slightly odd construction is necessary to create a Response object
+        # that has the correct URL already baked in (I think), because otherwise we get the following
+        # RuntimeError: Cannot call `raise_for_status` as the request instance has not been set on this response.
+        # TODO: find a better solution or understand the problem better
+        response._request = httpx.Request("GET", nonsense_url)
```

Comment:
I think this makes sense to me: I guess `raise_for_status` is dependent on a `Request` object being available to be able to generate an exception.

I think your solution looks fine, but 🍒 I would propose not replacing `response._request` post-hoc, and instead defining it where you define the `response` variable:

```python
response = httpx.Response(
    status_code=400, json={"key": "value"}, request=httpx.Request("GET", nonsense_url)
)
```

I find this easier to understand.

Reply:
Agree! Yours is much better